upcloud/resource_upcloud_server.go

```diff
@@ -420,94 +429,111 @@ func resourceUpCloudServerUpdate(ctx context.Context, d *schema.ResourceData, me
-			Address:     n.(string),
-			ServerUUID:  d.Id(),
-			StorageUUID: d.Get("template.0.id").(string),
-		}); err != nil {
+	// handle changes that need reboot
+	if d.HasChanges("plan", "cpu", "mem", "template", "storage_devices") {
+		if err := server.VerifyServerStopped(
+			request.StopServerRequest{
+				UUID: d.Id(),
+			},
+			meta,
+		); err != nil {
 			return diag.FromErr(err)
 		}
-	}
 
-	// handle the other storage devices
-	if d.HasChange("storage_devices") {
-		o, n := d.GetChange("storage_devices")
+		if plan, ok := d.GetOk("plan"); ok {
+			r.Plan = plan.(string)
+		} else {
+			r.CoreNumber = d.Get("cpu").(int)
+			r.MemoryAmount = d.Get("mem").(int)
+			r.Plan = "custom"
+		}
 
-		// detach the devices that should be detached or should be re-attached with different parameters
-		for _, storageDevice := range o.(*schema.Set).Difference(n.(*schema.Set)).List() {
-			if serverDetails.StorageDevice(storageDevice.(map[string]interface{})["storage"].(string)) == nil {
-				continue
+		// handle the template
+		if d.HasChanges("template.0.title", "template.0.size", "template.0.backup_rule") {
+			template := d.Get("template.0").(map[string]interface{})
+			if _, err := client.ModifyStorage(&request.ModifyStorageRequest{
+				UUID:  template["id"].(string),
+				Size:  template["size"].(int),
+				Title: template["title"].(string),
+				BackupRule: storage.BackupRule(
+					d.Get("template.0.backup_rule.0").(map[string]interface{}),
+				),
+			}); err != nil {
+				return diag.FromErr(err)
 			}
+		}
+		// should reattach if address changed
+		if d.HasChange("template.0.address") {
+			o, n := d.GetChange("template.0.address")
 			if _, err := client.DetachStorage(&request.DetachStorageRequest{
 				ServerUUID: d.Id(),
-				Address:    storageDevice.(map[string]interface{})["address"].(string),
+				Address:    o.(string),
 			}); err != nil {
 				return diag.FromErr(err)
 			}
-		}
-		// attach the storages that are new or have changed
-		for _, storageDevice := range n.(*schema.Set).Difference(o.(*schema.Set)).List() {
-			storageDevice := storageDevice.(map[string]interface{})
 			if _, err := client.AttachStorage(&request.AttachStorageRequest{
+				Address:     n.(string),
 				ServerUUID:  d.Id(),
-				Address:     storageDevice["address"].(string),
-				StorageUUID: storageDevice["storage"].(string),
-				Type:        storageDevice["type"].(string),
+				StorageUUID: d.Get("template.0.id").(string),
 			}); err != nil {
 				return diag.FromErr(err)
 			}
 		}
+
+		// handle the other storage devices
+		if d.HasChange("storage_devices") {
+			o, n := d.GetChange("storage_devices")
+
+			// detach the devices that should be detached or should be re-attached with different parameters
+			for _, storageDevice := range o.(*schema.Set).Difference(n.(*schema.Set)).List() {
+				if serverDetails.StorageDevice(storageDevice.(map[string]interface{})["storage"].(string)) == nil {
```

Comment:
Could we move this cast to a separate variable? It could make this a little bit readable with less `(map[string]interface{})`s all around.